Legacy/BootPlatform/BiosVideo/VideoBiosPatch.c

```diff
@@ -0,0 +1,325 @@
+/** @file
+  Copyright (C) 2020, Goldfish64. All rights reserved.
+
+  All rights reserved.
+
+  This program and the accompanying materials
+  are licensed and made available under the terms and conditions of the BSD License
+  which accompanies this distribution.  The full text of the license may be found at
+  http://opensource.org/licenses/bsd-license.php
+
+  THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,
+  WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
+**/
+
+#include "BiosVideo.h"
+
+STATIC
+BOOLEAN
+GetEdidMaxResolution (
+  IN  BIOS_VIDEO_DEV    *BiosVideoPrivate,
+  OUT UINT32            *Width,
+  OUT UINT32            *Height
+  )
+{
+  UINT8         Checksum;
+  UINT16        MaxWidth;
+  UINT16        MaxHeight;
+  UINT16        DtdWidth;
+  UINT16        DtdHeight;
+  UINT32        Index;
+
+  VESA_BIOS_EXTENSIONS_EDID_DATA_BLOCK  *Edid;
+  EDID_DTD                              *Dtds;
+
+  Edid = (VESA_BIOS_EXTENSIONS_EDID_DATA_BLOCK *) BiosVideoPrivate->EdidActive.Edid;
+  if (Edid == NULL) {
+    return FALSE;
+  }
+
+  //
+  // Ensure checksum matches.
+  //
+  Checksum = CalculateCheckSum8 (
+               (UINT8 *) Edid,
+               sizeof (VESA_BIOS_EXTENSIONS_EDID_DATA_BLOCK)
+               );
+  if (Checksum != 0) {
+    return FALSE;
+  }
+
+  Dtds      = (EDID_DTD *) Edid->DetailedTimingDescriptions;
+  MaxWidth  = 0;
+  MaxHeight = 0;
+
+  //
+  // Pull maximum supported resolution from detailed timing descriptors.
+  //
+  for (Index = 0; Index < VESA_BIOS_EXTENSIONS_DETAILED_TIMING_DESCRIPTOR_COUNT; Index++) {
+    DtdWidth  = Dtds[Index].HorzActivePixelsLsb | (Dtds[Index].HorzActivePixelsMsb << 8);
+    DtdHeight = Dtds[Index].VertActivePixelsLsb | (Dtds[Index].VertActivePixelsMsb << 8);
+
+    if (DtdWidth > MaxWidth && DtdHeight > MaxHeight) {
+      MaxWidth  = DtdWidth;
+      MaxHeight = DtdHeight;
+    }
+  }
+
+  if (MaxWidth == 0 || MaxHeight == 0) {
+    return FALSE;
+  }
+
+  *Width  = MaxWidth;
+  *Height = MaxHeight;
+  return TRUE;
+}
+
+STATIC
+VOID
+CalculateGtfTimings (
+  IN  UINT16      Width,
+  IN  UINT16      Height,
+  IN  UINT16      Frequency,
+  OUT UINT32      *ClockHz,
+  OUT UINT16      *HSyncStart,
+  OUT UINT16      *HSyncEnd,
+  OUT UINT16      *HBlank,
+  OUT UINT16      *VSyncStart,
+  OUT UINT16      *VSyncEnd,
+  OUT UINT16      *VBlank
+  )
+{
+  UINT32 Tmp1;
+  UINT32 Tmp2;
+  UINT32 Tmp3;
+
+  UINT32 Vbl;
+  UINT32 VFreq;
+  UINT32 Hbl;
+
+  Tmp1  = 11 * Frequency;
```

Comment:
Could you add any reference to an algorithm on how these values are calculated. Like a spec or a driver.